src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -3250,13 +3287,142 @@ internal virtual ImmutableArray<LabelSymbol> Labels
+                        break;
+                    }
+                    else
+                    {
+                        // the ones we found weren't viable. Try again, starting from the parent of where we found these
+                        patternType = patternType.GetNextBaseTypeNoUseSiteDiagnostics(null, Compilation, ref visitedTypes);
+                        lookupResult.Clear();
+                    }
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            MethodSymbol patternMethod = PerformPatternOverloadResolution(patternType, candidateMethods, syntaxExpr, warningsOnly, diagnostics, syntaxTree, messageID);
             candidateMethods.Free();
+            visitedTypes?.Free();
+
+            return patternMethod;
+        }
+
+        /// <summary>
+        /// Finds a pattern method implemented as an extension method, stopping at the first matching method
+        /// </summary>
+        /// <param name="patternType">Type to search.</param>
+        /// <param name="methodName">Method to search for.</param>
+        /// <param name="syntaxExpr">The expression for which lookup is being performed</param>
+        /// <param name="lookupResult">Passed in for reusability.</param>
+        /// <param name="diagnostics">Populated with binding diagnostics.</param>
+        /// <param name="messageID">The ID of the message to use when reporting diagnostics</param>
+        /// <returns>The desired method or null.</returns>
+        private MethodSymbol FindPatternExtensionMethod(TypeSymbol patternType, string methodName, LookupResult lookupResult,
+                                                        SyntaxNode syntaxExpr, DiagnosticBag diagnostics, MessageID messageID)
+        {
+            Debug.Assert(lookupResult.IsClear);
+
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            MethodSymbol patternMethod = null;
+
+            foreach (var scope in new ExtensionMethodScopes(this))
+            {
+                this.LookupExtensionMethodsInSingleBinder(scope, lookupResult, methodName, arity: 0, options: LookupOptions.Default, ref useSiteDiagnostics);
+
+                if (lookupResult.IsMultiViable)
+                {
+                    ArrayBuilder<MethodSymbol> candidateMethods = ArrayBuilder<MethodSymbol>.GetInstance();
+                    foreach (var symbol in lookupResult.Symbols)
+                    {
+                        //PROTOTYPE: for now, just check for the single this param. We probably want to also  allow e.g. dispose(this, int a = 4, params object[] args)
+                        if (symbol is MethodSymbol method && method.ParameterCount == 1)
+                        {
+                            candidateMethods.Add(method);
+                        }
+                    }
+
+                    if (candidateMethods.Count > 1)
```

Comment:
>if (candidateMethods.Count > 1) [](start = 20, length = 31)

Are we checking the `this` argument type? It seems reasonable to have multiple `Dispose` extension methods in the same scope, with different `this` arguments.

Reply:
Hmm, that's a great point. We're not, and the case you suggested fails. I've added the check that it's a valid receiver method and some tests to ensure we correctly handle that behaviour.